Automattic/woocommerce-payments
# Add WooPay Express Checkout functionality to shortcode and cart

#### Changes proposed in this Pull Request


As part of the effort to build a WooPay Express Checkout button, we have added the display of the button and the styling of the button. This PR is to add functionality to the buttons on the shortcode checkout and the cart pages. This does not impact the button block checkout or product page checkout. In short this PR implements the following changes:
- Display the WooPay iframe in a more generic centered position when triggered by the Express Checkout button.
- Allow the iframe to be triggered by a click on the React button compoenent.
- Pass the user's email address from the OTP verification to the session initialization.
- Redirect the user upon successful verification and session initialization.

This PR should be tested with https://github.com/Automattic/woopay/pull/1470 if it has not already been merged.

Some of the main implementation details:
- The iframe functionality was lifted and simplified from the existing `email-input-iframe.js` code. This simplification involved removing the `LoginSession` iframe flow, removing the popover arrow, and aligning the iframe with the email input field.
- The passing of the JS parameters proved to be quite tricky. There was already an existing `wcpayWooPayExpressParams` namespace that I ultimately ended up using. But what I found was that I needed a lot of the values that come from the [existing checkout fields](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payments-checkout.php#L95). The issue then was coming from the `getConfig` function pulling from the `wcpay_config` namespace. On the cart page I could define this JS namespace as `wcpay_config` and `getConfig` worked as expected. But on the checkout page `wcpay_config` was overwritten by the existing fields. So I ended up keeping the `wcpayWooPayExpressParams` namespace and then implementing a `getWooPayExpressData` function to be used in place of `getConfig` when necessary. This is copied from the `paymentRequest` model which I debated adding on to, but opted to create a parallel functionality instead.
- Passing of the email field from the OTP verification to the session initialization call rather than using the email input field from checkout (since we won't always have access to it from other pages).

**Known Issue**
I noticed that the `wcpayTracks` events were failing due to the same `getConfig` issue mentioned above. I'm breaking this out in to it's own [issue](https://github.com/Automattic/woopay/issues/1472).



#### Testing instructions



1. Be sure to be running WooPay with the https://github.com/Automattic/woopay/pull/1470 branch if it hasn't been merged yet.
2. Pull down this PR and compile JS.
3. Make sure the Express Checkout feature flag is enabled from the latest Dev Tools.
4. Make sure the Express Checkout button is set to display on Cart and Checkout page from Payments > Settings > WooPay Customize > Show express checkouts on.
5. Add a product to the cart.
6. Visit the cart, confirm display of WooPay button.
7. Click on WooPay button and confirm iframe loads with email input.
8. Enter email and verify OTP.
9. Confirm that session gets initialized and shopper is redirected to WooPay checkout.
10. Complete a full checkout and confirm successfully returned to merchant store with successful checkout.
11. Return to merchant store and add a product to the cart.
12. Visit the checkout page and confirm the display of the WooPay button.
13. Click on WooPay button and confirm iframe loads with email input.
14. Enter email and verify OTP.
15. Confirm that session gets initialized and shopper is redirected to WooPay checkout.
16. Complete a full checkout and confirm successfully returned to merchant store with successful checkout.

Would be good to also test a few other things:
- Confirm this is no regression on the existing WooPay checkout flow as these two paths should work in parallel while in development.
- Try out a few errors or failed login in the iframe and confirm all is handled properly on the merchant store.
- Try opening and closing the iframe with the button and make sure the DOM isn't getting cluttered and events are properly being unbound.



-------------------


**Post merge**
